src/_pytest/fixtures.py

```diff
@@ -1015,7 +1020,7 @@ def execute(self, request: SubRequest) -> FixtureValue:
             # note: comparison with `==` can fail (or be expensive) for e.g.
```

Comment:
TODO: The comment doesn't hold anymore. Numpy arrays would now be compared by parameter id if provided, or else by identity. Both methods won't fail and ain't expensive.

Reply:
> TODO: The comment doesn't hold anymore. Numpy arrays would now be compared by parameter id if provided, or else by identity. Both methods won't fail and ain't expensive.

Are you sure about this? IIUC, numpy arrays wrapped in `SafeHashWrapper` would still compare by `==`, so the expensive bit still holds.

Further, that makes me think that some types (like numpy arrays but also others) may implement `__eq__` but not `__hash__`. The `SafeHashWrapper` makes them implement `__hash__`, but the hash value doesn't match `__eq__` which is not allowed and can cause trouble. So this is a problem... Maybe `SafeHashWrapper` should check `Hashable` also in `__eq__`?

Reply:
>  numpy arrays wrapped in SafeHashWrapper would still compare by ==, so the expensive bit still holds

You're right, `numpy.ndarray.__eq__` is still executed and compares value by value. i.e. the expensive bit holds. Then it returns list (or exception) instead of bool which is catched by `SafeHashWrapper` and causes fallback to `id()`. I think the only reliable way to get rid of the expensive bit are user provided ids. If not given, we actually wish to prioritize fallback to deep value compare (possibly expensive) over fallback to id. That's the trade-off.

> the hash value doesn't match __eq__ which is not allowed and can cause trouble

Full ack. If `__eq__` and `__hash__` are not consistent yet, I'll have to fix it.

Reply:
> Maybe SafeHashWrapper should check Hashable also in __eq__?

Quite good idea, thanks, did that. As a bonus that makes at least numpy arrays cheap, because we now fall back early.

One remaining issue: A user could mutate parameters after reorder items took place, which renders the reordering invalid. It becomes obvious when we talk about hashability here, but I think the problem was always there, and I have no solution.